sky/backends/backend_utils.py

```diff
@@ -2010,14 +1662,42 @@ def check_owner_identity(cluster_name: str) -> None:
                 f'is {current_user_identity!r}.')
 
 
-def _get_cluster_status_via_cloud_cli(
+def _query_cluster_status_via_cloud_api(
     handle: 'cloud_vm_ray_backend.CloudVmRayResourceHandle'
-) -> List[global_user_state.ClusterStatus]:
+) -> List[status_lib.ClusterStatus]:
     """Returns the status of the cluster."""
-    resources: sky.Resources = handle.launched_resources
-    cloud = resources.cloud
+    cluster_name = handle.cluster_name
+    # Use region and zone from the cluster config, instead of the
+    # handle.launched_resources, because the latter may not be set
+    # correctly yet.
     ray_config = common_utils.read_yaml(handle.cluster_yaml)
-    return _QUERY_STATUS_FUNCS[str(cloud)](handle.cluster_name, ray_config)
+    provider_config = ray_config['provider']
+    region = provider_config.get('region') or provider_config.get('location')
+    zone = ray_config['provider'].get('availability_zone')
+    kwargs = {}
+    if isinstance(handle.launched_resources.cloud, clouds.GCP):
+        kwargs['use_tpu_vm'] = ray_config['provider'].get('_has_tpus', False)
+
+    # Query the cloud provider.
+    node_statuses = handle.launched_resources.cloud.query_status(
+        cluster_name, {'ray-cluster-name': cluster_name}, region, zone,
+        **kwargs)
+    # GCP does not clean up preempted TPU VMs. We remove it ourselves.
+    # TODO(wei-lin): handle multi-node cases.
+    if kwargs.get('use_tpu_vm', False) and len(node_statuses) == 0:
+        logger.debug(f'Terminating preempted TPU VM cluster {cluster_name}')
+        backend = backends.CloudVmRayBackend()
+        # Do not use refresh cluster status during teardown, as that will
+        # cause inifinite recursion by calling cluster status refresh
+        # again.
+        # Post teardown cleanup be done later in this function, which will remove
+        # the cluster entry from the status table & the ssh config file.
+        backend.teardown_no_lock(handle,
+                                 terminate=True,
+                                 purge=False,
+                                 post_teardown_cleanup=False,
+                                 refresh_cluster_status=False)
```

Comment:
While we're doing this refactor, should we try to move this functionality (terminating preempted TPU VMs) a little up in the stack? This method should likely only query the cloud API and this TPU-specific handling can perhaps be done in `_update_cluster_status_no_lock`, where we have other TPU-specific code too.

Reply:
Actually, I think this should be handled within the `GCP.query_status()` function, once we have refactored the termination code path into the cloud class. Reason: the TPU VM will stay in PREEMPTED status and holds the quota, even if that cluster is completely useless and should be cleaned up immediately. 
I tend to leave it here, and once we refactored the termination code path, we can move this into the `GCP.query_status()`. Added acomment about this. Wdyt?